brave/brave-ios
# Fix #6367: Send NFTs

## Summary of Changes
- Add support for sending ERC721 NFTs on Ethereum / EVMs
- Add support for sending Solana NFTs on Solana networks
- Resolve race condition in Send view when switching networks

This pull request fixes #6367

## Submitter Checklist:


## Test Plan:
Ethereum / ERC721 NFT:
1. Add at least one ERC721 asset you do not own, and one you do own (needs balance)
2. Open Send view, enter a valid ethereum address select the ERC721 NFT asset you do not own
3. Verify the send button is disabled (0 balance)
4. Select the ERC721 NFT asset you do own
5. Verify the send button is enabled, tap send
6. Verify transaction is created and displayed, expected gas balance shown
7. Send the transaction, verify transaction displays in Account Details list

Solana NFT:
1. Add at least one Solana NFT asset you do not own, and one you do own (needs balance)
2. Open Send view, enter a valid Solana address select the Solana NFT asset you do not own
3. Verify the send button is disabled (0 balance)
4. Select the Solana NFT asset you do own
5. Verify the send button is enabled, tap send
6. Verify transaction is created and displayed, expected gas balance shown
7. Send the transaction, verify transaction displays in Account Details list

Insufficient Balance:
1. Open send and enter a valid address for the network type
2. Tap 100%
3. Verify 'send' button is enabled
4. Add any non-zero digit
5. Verify 'send' button becomes 'Insufficient balance'
6. Select an NFT with balance
7. Verify 'send' button is enabled
8. Select an NFT without any balance
9. Verify 'send' button is disabled

## Screenshots:

Ethereum ERC721 NFT:

https://user-images.githubusercontent.com/5314553/201737660-9f5fd8a9-7370-4f31-b535-51a8ddff4c2f.mp4 

Solana NFT:

https://user-images.githubusercontent.com/5314553/201744224-206bbb6c-3050-4c1a-939b-5bf9ba0dfef5.mp4

Insufficient Balance:

https://user-images.githubusercontent.com/5314553/201953463-2c82bf20-74c8-4e04-96ab-889320814cb0.mp4


## Reviewer Checklist:

  - `QA/(Yes|No)`
  - `bug` / `enhancement`
